awx/main/models/inventory.py

```diff
@@ -1797,3 +1815,651 @@ class Meta:
+            elif k == 'satellite6_group_prefix' and isinstance(v, str):
+                group_prefix = v
+            elif k == 'satellite6_want_hostcollections' and isinstance(v, bool):
+                want_hostcollections = v
+            else:
+                cp.set(section, k, str(v))
+
+        if credential:
+            cp.set(section, 'url', credential.get_input('host', default=''))
+            cp.set(section, 'user', credential.get_input('username', default=''))
+            cp.set(section, 'password', credential.get_input('password', default=''))
+
+        section = 'ansible'
+        cp.add_section(section)
+        cp.set(section, 'group_patterns', group_patterns)
+        cp.set(section, 'want_facts', 'True')
+        cp.set(section, 'want_hostcollections', str(want_hostcollections))
+        cp.set(section, 'group_prefix', group_prefix)
+
+        section = 'cache'
+        cp.add_section(section)
+        cp.set(section, 'path', '/tmp')
+        cp.set(section, 'max_age', '0')
+
+        return self.dump_cp(cp, credential)
+
+
+class cloudforms(PluginFileInjector):
+    ini_env_reference = 'CLOUDFORMS_INI_PATH'
+    # Also no base_injector because this does not work in playbooks
+
+    def build_script_private_data(self, inventory_update, private_data_dir):
+        cp = configparser.RawConfigParser()
+        credential = inventory_update.get_cloud_credential()
+
+        section = 'cloudforms'
+        cp.add_section(section)
+
+        if credential:
+            cp.set(section, 'url', credential.get_input('host', default=''))
+            cp.set(section, 'username', credential.get_input('username', default=''))
+            cp.set(section, 'password', credential.get_input('password', default=''))
+            cp.set(section, 'ssl_verify', "false")
+
+        cloudforms_opts = dict(inventory_update.source_vars_dict.items())
+        for opt in ['version', 'purge_actions', 'clean_group_keys', 'nest_tags', 'suffix', 'prefer_ipv4']:
+            if opt in cloudforms_opts:
+                cp.set(section, opt, str(cloudforms_opts[opt]))
+
+        section = 'cache'
+        cp.add_section(section)
+        cp.set(section, 'max_age', "0")
+        cache_path = tempfile.mkdtemp(
+            prefix='cloudforms_cache',
+            dir=private_data_dir
+        )
+        cp.set(section, 'path', cache_path)
+
+        return self.dump_cp(cp, credential)
+
+
+class tower(PluginFileInjector):
+    base_injector = 'template'
+
+    def get_script_env(self, inventory_update, private_data_dir, private_data_files):
+        env = super(tower, self).get_script_env(inventory_update, private_data_dir, private_data_files)
+        env['TOWER_INVENTORY'] = inventory_update.instance_filters
+        env['TOWER_LICENSE_TYPE'] = get_licenser().validate().get('license_type', 'unlicensed')
+        return env
+
+
+for cls in PluginFileInjector.__subclasses__():
+    InventorySourceOptions.injectors[cls.__name__] = cls
```

Comment:
I _really_ like this pattern.  Long term, I think an awesome thing to do here would be to unit test _these_ injection implementations and lose all the crazy mock-based tests in `tests/unit/test_tasks`.

Reply:
> I _really_ like this pattern.

well, it is a pattern copied from credentials. The semantics can be implemented several ways, but that's easy to change. For native credential types you added them to the tracking dict in `__init__` method (the PR you _just_ merged), which might be the only other option I liked better than the `__subclasses__` method here.

> I think an awesome thing to do here would be to unit test _these_ injection implementations and lose all the crazy mock-based tests in `tests/unit/test_tasks`.

I've been talking with @chrismeyersfsu a lot about the `build_private_data` and `build_private_data_files` methods. The way we pass data between the two is complicated and confusing, but I _really_ like the idea of first collecting what we're going to write to disk, and then having a final step where we create the directory & files. In both tests_tasks.py and the new content here, we have to write things to the filesystem, then pull them _back out_, rearrange them, and make assertions. Having the boundary between generating the content and writing it to disk could make that much more enjoyable.